### PR TITLE
(port) Crystalized Crossbreeds (Adamantine) and ports bull terriers

### PR DESCRIPTION
**Pull request by boodaliboo**
original pr https://github.com/BeeStation/BeeStation-Hornet/pull/2946
original pr 2  https://github.com/tgstation/tgstation/pull/53148
"Dominant Adamantine Extracts (Crystalized Extract) provide small area effects that last until the crystal is destroyed. Activate in hand to "build" the crystal in place. Crystals are immovable and solid, but may be broken with one strike from anything but bare hands
If any crystal is within another's area of effect, both crystals shatter, preventing their effects from being stacked
Grey - Slowly feeds all slimes in the area, effect blocked by walls
Orange - Ignites anyone in range of the crystal, heats surrounding air to 500k. Effect blocked by walls
Purple - Restores 1 of a random damage type to everything in range every time it ticks, including brain and cell damage
Blue - Stabilizes atmos in effect range to the standard O2/N2 mixture, and stabilizes temperature to 300k. Effect blocked by walls
Metal - just like Purple but affects only borgs
Yellow - Hit it with a cell to instantly charge it, cells that are already fully charged explode.
Dark Purple - Consumes plasma in the air, converting it into plasma sheets. Crystal releases and ignites a small amount of plasma when destroyed
Dark Blue - Cleans and dries tiles in the area
Silver - Plants grow slightly faster, and prevents pests from growing at all.
Bluespace - Acts as a beacon to other crystals of this type. Click with an empty hand to teleport between them
Sepia - Everything in the area is under the effect simmiliar to Stasis (Real stasis is a hardstun)
Cerulean - Adjacent tiles will gradually grow slime crystals, up to a maximum stack of 5 in each tile. slime crystals may be combined with any material sheet to increase its stack size by 5.
Pyrite - Causes floor tiles to be randomly colored within area of effect
Red - Crystal cleans blood from the ground in a wide radius, and may store up to 300u of gathered blood. Crystal may be interacted with to consume some blood and create a piece of "meat" or a random, functional organ. Containers may also be used on the crystal to transfer blood directly out of it.
Green - Crystal stores one random mutation from the last player who interacted with it (starts with none). Players standing within the effect radius are gradually cured of mutations, but will be forcefully given the mutation stored in the crystal
Pink - Everyone within range of the crystal is pacified (effects identical to pacifist trait).
Gold - Click crystal with empty hand to be transformed into a random pet-type simplemob. Leave crystal's area of effect to revert
Oil - Lubes tiles within area of effect.
Black - Players standing within two tiles of the crystal will begin slowly transforming into a random slimeperson. Has no effect on slimepeople.
Light Pink - Ghosts may interact with the crystal to become pink lightgeists that can't leave the crystal's area of effect. They may not speak but are still able to heal.
Adamantine - Makes everyone slightly more resistant in an area around it.
Rainbow - Does nothing on its own, but will accept Crystalized extracts to begin emitting the same effect as the consumed extract, allowing for multiple different effects in the same space

Why It's Good For The Game
More crossbreeds = good"
#### Changelog

:cl:  
rscadd: Added new crossbreeds

/:cl:

### CODE REVIEW

**Comment by boodaliboo** [comment]
for some reason i needed bull terriers for this pr

**Comment by Tipy1802** [comment]
Wait for the ''hurr durr xenobio still exists and this PR does not nerf it to the ground'' crowd to appear

**Comment by boodaliboo** [comment]
> Wait for the ''hurr durr xenobio still exists and this PR does not nerf it to the ground'' crowd to appear

xenobio is so much worse then it used to be especially because of the stable extract nerf

**Comment by Tipy1802** [comment]
Agreed people who scape goat xenobio are a clowns at this point. There hasn't been an issue of xenobio using OP stuff to validhunt in a long ass time. It's usually either mining or engineering that validhunts scientists just keep to themself and upgrade the machines every so often

**Comment by Mqiib** [comment]
This is cool but have you tested it?

**Comment by boodaliboo** [comment]
> This is cool but have you tested it?

my dream maker wont compile yogs but i did go on a bee branch and fucked around with the crossbreeds before making this

**Review by fluffe9911 — APPROVED** [review]
really hope I dont end up regretting this but on the surface it seems fine

**Comment by Marmio64** [comment]
feel like the yellow one is going to be used with the chaplain in order to mass produce androids but whether or not it is a bad thing is up to the individual

**Comment by ghost** [comment]
Merging this and making xenobiology any stronger than the godawful overpowered monster it already is is a horrific idea.

File: code/modules/cargo/packs.dm [hunk 1]
```diff
@@ -2517,3 +2525,5 @@
 					/obj/item/toner,
 					/obj/item/toner)
 	crate_name = "toner crate"
+
+
```
Comment (JamieD1) [inline code/modules/cargo/packs.dm]:
```suggestion
```

File: code/modules/research/xenobiology/crossbreeding/_misc.dm [hunk 2]
```diff
@@ -310,3 +310,27 @@ Slimecrossing Items
...
+	if(istype(stack_item,/obj/item/stack/telecrystal))
+		var/mob/living/carbon/carbie = user
+		to_chat(user,"<span class='big red'>You will pay for your hubris!</span>")
+		carbie.gain_trauma(/datum/brain_trauma/special/beepsky,TRAUMA_RESILIENCE_ABSOLUTE)
+		qdel(src)
+		return
+
+	stack_item.add(amt)
+
+	qdel(src)
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_misc.dm]:
```suggestion
	stack_item.add(amt)
	qdel(src)
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+	if(master_crystal)
+		invisibility = INVISIBILITY_MAXIMUM
+		max_integrity = 1000
+		obj_integrity = 1000
+
+/obj/structure/slime_crystal/Initialize()
+	. = ..()
+	name =  "[colour] slimic pylon"
+	var/itemcolor = "#FFFFFF"
+
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 4]
```diff
@@ -0,0 +1,627 @@
...
+	if(uses_process)
+		STOP_PROCESSING(SSobj, src)
+	for(var/X in affected_mobs)
+		on_mob_leave(X)
+	return ..()
+
+/obj/structure/slime_crystal/process()
+	if(!uses_process)
+		return PROCESS_KILL
+
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 5]
```diff
@@ -0,0 +1,627 @@
...
+	var/list/current_mobs = view_or_range(3, src, range_type)
+	for(var/mob/living/mob_in_range in current_mobs)
+		if(!(mob_in_range in affected_mobs))
+			on_mob_enter(mob_in_range)
+			affected_mobs[mob_in_range] = 0
+
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 6]
```diff
@@ -0,0 +1,627 @@
...
+		affected_mobs[mob_in_range]++
+		on_mob_effect(mob_in_range)
+
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+	var/turf/open/T = get_turf(src)
+	if(!istype(T))
+		return
+	var/datum/gas_mixture/gas = T.return_air()
+	gas.set_temperature(T0C + 200)
+	T.air_update_turf()
+
+/obj/structure/slime_crystal/purple
+	colour = "purple"
+
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 8]
```diff
@@ -0,0 +1,627 @@
...
+	var/heal_amt = 2
+
+/obj/structure/slime_crystal/purple/on_mob_effect(mob/living/affected_mob)
+	if(!istype(affected_mob, /mob/living/carbon))
+		return
+	var/mob/living/carbon/carbon_mob = affected_mob
+	var/rand_dam_type = rand(0, 10)
+
+	new /obj/effect/temp_visual/heal(get_turf(affected_mob), "#e180ff")
+
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
	new /obj/effect/temp_visual/heal(get_turf(affected_mob), "#e180ff")
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 9]
```diff
@@ -0,0 +1,627 @@
...
+		if(isspaceturf(T))
+			continue
+		var/datum/gas_mixture/gas = T.return_air()
+		gas.parse_gas_string(OPENTURF_DEFAULT_ATMOS)
+		T.air_update_turf()
+
+/obj/structure/slime_crystal/metal
+	colour = "metal"
+
+	var/heal_amt = 1
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
	var/heal_amt = 1
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 10]
```diff
@@ -0,0 +1,627 @@
...
+		if(cell.charge == cell.maxcharge)
+			to_chat("<span class = 'danger'> You try to charge the cell, but it is already fully energized. You are not sure if this was a good idea...")
+			cell.explode()
+			return
+		to_chat("<span class = 'notice'> You charged the [I.name] on [name]!")
+		cell.give(cell.maxcharge)
+		return
+	return ..()
+/obj/structure/slime_crystal/darkpurple
+	colour = "dark purple"
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
	return ..()

/obj/structure/slime_crystal/darkpurple
	colour = "dark purple"
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 11]
```diff
@@ -0,0 +1,627 @@
...
+
+/obj/structure/slime_crystal/darkpurple/process()
+	var/turf/T = get_turf(src)
+	if(!istype(T, /turf/open))
+		return
+	var/turf/open/open_turf = T
+	var/datum/gas_mixture/air = open_turf.return_air()
+
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 12]
```diff
@@ -0,0 +1,627 @@
...
+	colour = "dark blue"
+
+/obj/structure/slime_crystal/darkblue/process()
+	var/list/listie = range(5,src)
+	for(var/turf/open/T in listie)
+		if(prob(75))
+			continue
+		var/turf/open/open_turf = T
+		open_turf.MakeDry(TURF_WET_LUBE)
+
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 13]
```diff
@@ -0,0 +1,627 @@
...
+		assoc_list[name] = BSC
+
+	var/chosen_input = input(user,"What destination do you want to choose",null) as null|anything in assoc_list
+	in_use = FALSE
+
+	if(!chosen_input || !assoc_list[chosen_input])
+		return
+
+	do_teleport(user ,assoc_list[chosen_input])
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
/obj/structure/slime_crystal/bluespace/attack_hand(mob/user)
	if(in_use)
		return
	var/list/local_bs_list = GLOB.bluespace_slime_crystals.Copy()
	local_bs_list -= src
	if(!LAZYLEN(local_bs_list))
		return ..()
	if(local_bs_list.len == 1)
		do_teleport(user, local_bs_list[1])
		return
	in_use = TRUE
	var/list/assoc_list = list()
	for(var/BSC in local_bs_list)
		var/area/bsc_area = get_area(BSC)
		var/name = "[bsc_area.name] bluespace slimic pylon"
		var/counter = 0
		do
			counter++
		while(assoc_list["[name]([counter])"])
		name += "([counter])"
		assoc_list[name] = BSC
	var/chosen_input = input(user,"What destination do you want to choose",null) as null|anything in assoc_list
	in_use = FALSE
	if(!chosen_input || !assoc_list[chosen_input])
		return
	do_teleport(user ,assoc_list[chosen_input])
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 14]
```diff
@@ -0,0 +1,627 @@
...
+		density = TRUE
+
+	stage ++
+
+	var/matrix/M = new
+	M.Scale(1/max_stage * stage)
+
+	animate(src, transform = M, time = 60 SECONDS)
+
+	addtimer(CALLBACK(src, .proc/stage_growth), 60 SECONDS)
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
/obj/structure/cerulean_slime_crystal/proc/stage_growth()
	if(stage == max_stage)
		return
	if(stage == 3)
		density = TRUE
	stage ++
	var/matrix/M = new
	M.Scale(1/max_stage * stage)
	animate(src, transform = M, time = 60 SECONDS)
	addtimer(CALLBACK(src, .proc/stage_growth), 60 SECONDS)
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 15]
```diff
@@ -0,0 +1,627 @@
...
+
+	addtimer(CALLBACK(src,.proc/change_colour),rand(0.75 SECONDS,1.25 SECONDS))
+
+/obj/structure/slime_crystal/red
+	colour = "red"
+
+	var/blood_amt = 0
+
+	var/max_blood_amt = 300
+
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
		T.add_atom_colour(pick(color_list), FIXED_COLOUR_PRIORITY)
	addtimer(CALLBACK(src,.proc/change_colour),rand(0.75 SECONDS,1.25 SECONDS))

/obj/structure/slime_crystal/red
	colour = "red"
	var/blood_amt = 0
	var/max_blood_amt = 300
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 16]
```diff
@@ -0,0 +1,627 @@
...
+		blood_amt++
+		if(blood_amt == max_blood_amt)
+			return
+
+/obj/structure/slime_crystal/red/attack_hand(mob/user)
+	if(blood_amt < 100)
+		return ..()
+
+	blood_amt -= 100
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
/obj/structure/slime_crystal/red/process()
	if(blood_amt == max_blood_amt)
		return
	for(var/obj/effect/decal/cleanable/blood/B in range(3,src))
		qdel(B)
		blood_amt++
		if(blood_amt == max_blood_amt)
			return

/obj/structure/slime_crystal/red/attack_hand(mob/user)
	if(blood_amt < 100)
		return ..()
	blood_amt -= 100
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 17]
```diff
@@ -0,0 +1,627 @@
...
+	if(!istype(I, /obj/item/reagent_containers/glass/beaker))
+		return ..()
+
+	var/obj/item/reagent_containers/glass/beaker/item_beaker = I
+
+	if(!item_beaker.is_refillable() || (item_beaker.reagents.total_volume + 10 > item_beaker.reagents.maximum_volume))
+		return ..()
+	blood_amt -= 10
+	item_beaker.reagents.add_reagent(/datum/reagent/blood,10)
+
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
/obj/structure/slime_crystal/red/attacked_by(obj/item/I, mob/living/user)
	if(blood_amt < 10)
		return ..()
	if(!istype(I, /obj/item/reagent_containers/glass/beaker))
		return ..()
	var/obj/item/reagent_containers/glass/beaker/item_beaker = I
	if(!item_beaker.is_refillable() || (item_beaker.reagents.total_volume + 10 > item_beaker.reagents.maximum_volume))
		return ..()
	blood_amt -= 10
	item_beaker.reagents.add_reagent(/datum/reagent/blood,10)
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 18]
```diff
@@ -0,0 +1,627 @@
...
+	var/list/secondary_list = list()
+
+	for(var/X in mut_list)
+		if(istype(X,stored_mutation))
+			continue
+		var/datum/mutation/t_mutation = X
+		secondary_list += t_mutation.type
+
+	var/datum/mutation/mutation = pick(secondary_list)
+	human_mob.dna.remove_mutation(mutation)
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
/obj/structure/slime_crystal/green/on_mob_effect(mob/living/affected_mob)
	if(!ishuman(affected_mob) || !stored_mutation || HAS_TRAIT(affected_mob,TRAIT_BADDNA))
		return
	var/mob/living/carbon/human/human_mob = affected_mob
	human_mob.dna.add_mutation(stored_mutation)
	if(affected_mobs[affected_mob] % 60 != 0)
		return
	var/list/mut_list = human_mob.dna.mutations
	var/list/secondary_list = list()
	for(var/X in mut_list)
		if(istype(X,stored_mutation))
			continue
		var/datum/mutation/t_mutation = X
		secondary_list += t_mutation.type
	var/datum/mutation/mutation = pick(secondary_list)
	human_mob.dna.remove_mutation(mutation)
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 19]
```diff
@@ -0,0 +1,627 @@
...
+/obj/structure/slime_crystal/gold/on_mob_leave(mob/living/affected_mob)
+	if(!istype(affected_mob,/mob/living/simple_animal/pet))
+		return
+
+	var/mob/living/carbon/human/human_mob = locate() in affected_mob
+
+	if(!human_mob)
+		return
+
+	affected_mob.mind.transfer_to(human_mob)
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
/obj/structure/slime_crystal/gold/on_mob_leave(mob/living/affected_mob)
	if(!istype(affected_mob,/mob/living/simple_animal/pet))
		return
	var/mob/living/carbon/human/human_mob = locate() in affected_mob
	if(!human_mob)
		return
	affected_mob.mind.transfer_to(human_mob)
```

File: code/modules/research/xenobiology/crossbreeding/_structures.dm [hunk 20]
```diff
@@ -0,0 +1,627 @@
...
+
+	if(!istype(I,/obj/item/slimecross/crystalized) || istype(I,/obj/item/slimecross/crystalized/rainbow))
+		return
+
+	var/obj/item/slimecross/crystalized/slimecross = I
+
+	if(inserted_cores[slimecross.type])
+		return
+
+	inserted_cores[slimecross.type] = new slimecross.crystal_type(get_turf(src),src)
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/_structures.dm]:
```suggestion
	var/mob/living/carbon/human/human = affected_mob
	human.dna.species.brutemod += 0.1
	human.dna.species.burnmod += 0.1

/obj/structure/slime_crystal/rainbow
	colour = "rainbow"
	uses_process = FALSE
	var/list/inserted_cores = list()

/obj/structure/slime_crystal/rainbow/Initialize()
	. = ..()
	for(var/X in subtypesof(/obj/item/slimecross/crystalized) - /obj/item/slimecross/crystalized/rainbow)
		inserted_cores[X] = FALSE

/obj/structure/slime_crystal/rainbow/attacked_by(obj/item/I, mob/living/user)
	. = ..()
	if(!istype(I,/obj/item/slimecross/crystalized) || istype(I,/obj/item/slimecross/crystalized/rainbow))
		return
	var/obj/item/slimecross/crystalized/slimecross = I
	if(inserted_cores[slimecross.type])
		return
	inserted_cores[slimecross.type] = new slimecross.crystal_type(get_turf(src),src)
```

File: code/modules/research/xenobiology/crossbreeding/crystalized.dm [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+	if(C)
+		to_chat(user,"<span class='notice'>You can't build crystals that close to each other!</span>")
+		return
+
+	var/user_turf = get_turf(user)
+
+	if(!do_after(user,15 SECONDS,FALSE,user_turf))
+		return
+
+	new crystal_type(user_turf)
```
Comment (JamieD1) [inline code/modules/research/xenobiology/crossbreeding/crystalized.dm]:
```suggestion
	var/obj/structure/slime_crystal/C = locate() in range(6,get_turf(user))
	if(C)
		to_chat(user,"<span class='notice'>You can't build crystals that close to each other!</span>")
		return
	var/user_turf = get_turf(user)
	if(!do_after(user,15 SECONDS,FALSE,user_turf))
		return
	new crystal_type(user_turf)
```

**Review by JamieD1 — APPROVED** [review]

**Pull request merged**